src/ble.rs

```diff
@@ -0,0 +1,179 @@
+//! Connect to an LED badge via Bluetooth Low Energy (BLE)
+
+use std::time::Duration;
+
+use anyhow::{Context, Result};
+use btleplug::{
+    api::{Central as _, Manager as _, Peripheral as _, ScanFilter, WriteType},
+    platform::{Manager, Peripheral},
+};
+use tokio::time;
+use uuid::Uuid;
+
+use crate::protocol::PayloadBuffer;
+
+/// `0000fee0-0000-1000-8000-00805f9b34fb`
+const BADGE_SERVICE_UUID: Uuid = btleplug::api::bleuuid::uuid_from_u16(0xfee0);
+/// `0000fee1-0000-1000-8000-00805f9b34fb`
+const BADGE_CHAR_UUID: Uuid = btleplug::api::bleuuid::uuid_from_u16(0xfee1);
+
+const BADGE_BLE_DEVICE_NAME: &str = "LSLED";
+const BLE_CHAR_CHUNK_SIZE: usize = 16;
+
+/// A discovered BLE device
+pub struct Device {
+    peripheral: Peripheral,
+}
+
+impl Device {
+    /// Return all supported devices that are found in two seconds.
+    ///
+    /// Returns all badges that are in BLE range and are in Bluetooth transfer mode.
+    pub async fn enumerate() -> Result<Vec<Self>> {
+        Self::enumerate_duration(Duration::from_secs(2)).await
+    }
+
+    /// Return all supported devices that are found in the given duration.
+    ///
+    /// Returns all badges that are in BLE range and are in Bluetooth transfer mode.
+    /// # Panics
+    /// This function panics if it is unable to access the Bluetooth adapter.
+    pub async fn enumerate_duration(scan_duration: Duration) -> Result<Vec<Self>> {
+        // Run device scan
+        let manager = Manager::new().await.context("create BLE manager")?;
+        let adapters = manager
+            .adapters()
+            .await
+            .context("enumerate bluetooth adapters")?;
+        let adapter = adapters.first().context("no bluetooth adapter found")?;
+
+        adapter
+            .start_scan(ScanFilter {
+                services: vec![BADGE_SERVICE_UUID],
+            })
+            .await
+            .context("bluetooth scan start")?;
+        time::sleep(scan_duration).await;
+
+        // Filter for badge devices
+        let mut led_badges = vec![];
+        for p in adapter
+            .peripherals()
+            .await
+            .context("enumerating bluetooth devices")?
+        {
+            if let Some(badge) = Self::from_peripheral(p).await {
+                led_badges.push(badge);
+            }
+        }
+
+        Ok(led_badges)
+    }
+
+    async fn from_peripheral(peripheral: Peripheral) -> Option<Self> {
+        // Check whether the BLE device has the service UUID we're looking for
+        // and also the correct name.
+        // The service uuid is also by devices that are not LED badges, so
+        // the name check is also necessary.
+        let props = peripheral.properties().await;
+        if props.is_err() {
+            return None;
+        }
+
+        if let Some(props) = props.unwrap() {
```

Comment:
Let's eliminate the last two unwraps. Unwrap 1/2:
```suggestion
        let props = peripheral.properties().await.ok()??;
```

The first `.ok()?` converts the `Result` into an `Option` and returns `None` if there was an `Err`. The second `?` returns `None` if the inner `Option` was `None`.